core/aggregations/src/main/java/com/microsoft/gctoolkit/aggregations/RuntimeAggregation.java

```diff
@@ -0,0 +1,75 @@
+package com.microsoft.gctoolkit.aggregations;
+
+import com.microsoft.gctoolkit.aggregator.Aggregation;
+import com.microsoft.gctoolkit.time.DateTimeStamp;
+
+/**
+ * An Aggregation that collates runtime data. This class is meant to be extended by other
+ * implementations that need access to runtime data to perform calculations. For example,
+ * to calculate the ratio of pause time to runtime duration.
+ */
+public abstract class RuntimeAggregation  implements Aggregation {
+
+    private volatile DateTimeStamp timeOfFirstEvent = null;
+    private volatile DateTimeStamp timeOfLastEvent = new DateTimeStamp(0d);
+
+    /**
+     * Runtime duration (in decimal seconds) after which we no longer consider the GC log to be a fragment
+     */
+    private static final double LOG_FRAGMENT_THRESHOLD = 18d;
+
+    /**
+     * This class is meant to be extended.
+     */
+    protected RuntimeAggregation() {}
+
+    /**
+     * RuntimeAggregation collates the time of an event and the duration of the event.
+     * @param eventTime The time a JVMEvent occurred.
+     * @param duration The duration of the JVMEvent.
+     */
+    public void record(DateTimeStamp eventTime, double duration) {
+        duration = !Double.isNaN(duration) ? duration : 0d;
+
+        if (timeOfFirstEvent == null || (eventTime != null && eventTime.before(timeOfFirstEvent))) {
+            timeOfFirstEvent = eventTime != null ? eventTime : new DateTimeStamp(0d);
+        }
+
+        final DateTimeStamp now =
+                eventTime != null ? eventTime.add(duration) : timeOfLastEvent.add(duration);
+        if (now.after(timeOfLastEvent)) {
+            timeOfLastEvent = now;
```

Comment:
In the case of clock rollbacks in the log, it's best to let the timeOfLastEvent take on the new value. This will likely result in a corrupted calculation, maybe 2.. but after that if there are no rollbacks, then all off the subsequent calculations of now should be ok

Reply:
I'll look into this.

Reply:
In Censum, timeOfLastEvent is calculated as given.
```
            vertx.eventBus().
                    consumer(mailBox, message -> {
                        try {
                            JVMEvent event = (JVMEvent) message.body();
                            if (event instanceof JVMTermination)
                                return;
                            DateTimeStamp now = event.getDateTimeStamp().add(event.getDuration());
                            if (now.after(timeOfLastEvent)) {
                                timeOfLastEvent = now;
                            }
                        } catch (Throwable t) {
                            LOGGER.throwing(this.getClass().getName(), "start", t);
                        }
                    });

```